pkg/pillar/cmd/zedrouter/probe.go

```diff
@@ -0,0 +1,538 @@
+		for _, st := range items {
+			netstatus := cast.CastNetworkInstanceStatus(st)
+			if !isSharedPortLabel(netstatus.Port) && netstatus.Port != port.Name {
+				continue
+			}
+			niProbingUpdatePort(ctx, port, &netstatus)
+			checkNIprobeUplink(ctx, &netstatus)
+		}
+	}
+}
+
+// called from handleNetworkInstanceModify and handleNetworkInstanceCreate
+func niUpdateNIprobing(ctx *zedrouterContext, status *types.NetworkInstanceStatus) {
+	pub := ctx.subDeviceNetworkStatus
+	items := pub.GetAll()
+	portList := getIfNameListForPort(ctx, status.Port)
+	log.Infof("niUpdateNIprobing: enter, number of ports %d\n", len(portList))
+	for _, st := range items {
+		devStatus := cast.CastDeviceNetworkStatus(st)
+
+		for _, port := range portList {
+			devPort := getDevPort(&devStatus, port)
+			if devPort == nil {
+				continue
+			}
+			for _, ipinfo := range devPort.AddrInfoList {
+				log.Infof("niUpdateNIprobing: port %s, free %v, addr %v, Gw %v\n",
+					devPort.IfName, devPort.Free, ipinfo.Addr, devPort.Gateway)
+			}
+			if !isSharedPortLabel(status.Port) && status.Port != devPort.Name {
+				continue
+			}
+			niProbingUpdatePort(ctx, *devPort, status)
+		}
+	}
+	checkNIprobeUplink(ctx, status)
+}
+
+func getDevPort(status *types.DeviceNetworkStatus, port string) *types.NetworkPortStatus {
+	for _, tmpport := range status.Ports {
+		if strings.Compare(tmpport.Name, port) == 0 {
+			return &tmpport
+		}
+	}
+	return nil
+}
+
+func niProbingUpdatePort(ctx *zedrouterContext, port types.NetworkPortStatus,
+	netstatus *types.NetworkInstanceStatus) {
+	log.Infof("niProbingUpdatePort: enter\n")
+	if _, ok := netstatus.PInfo[port.IfName]; !ok {
+		if port.IfName == "" { // no need to probe for air-gap type of NI
+			return
+		}
+		info := types.ProbeInfo{
+			IfName:       port.IfName,
+			IsPresent:    true,
+			GatewayUP:    true,
+			NhAddr:       port.Gateway,
+			LocalAddr:    portGetIntfAddr(port),
+			Class:        getIntfClassByIOBundle(ctx, port),
+			RemoteHostUP: true,
+		}
+		netstatus.PInfo[port.IfName] = info
+		log.Infof("niProbingUpdatePort: %s assigned new %s, info len %d, class %v\n",
+			netstatus.BridgeName, port.IfName, len(netstatus.PInfo), info.Class)
+	} else {
+		info := netstatus.PInfo[port.IfName]
+		info.IsPresent = true
+		info.NhAddr = port.Gateway
+		info.LocalAddr = portGetIntfAddr(port)
+		info.FailedCnt = 0
+		info.SuccessCnt = 0
+		netstatus.PInfo[port.IfName] = info
+		log.Infof("niProbingUpdatePort: %s modified %s", netstatus.BridgeName, port.IfName)
+	}
+	publishNetworkInstanceStatus(ctx, netstatus)
+}
+
+func getIntfClassByIOBundle(ctx *zedrouterContext, port types.NetworkPortStatus) types.IntfClass {
+	aa := ctx.assignableAdapters
+	if aa == nil { // no information
+		log.Infof("getIntfClassIOBundle: aa is nil")
+		return types.ClassETHER
+	}
+	ioBundle := aa.LookupIoBundleNet(port.IfName)
+	if ioBundle == nil {
+		log.Infof("getIntfClassIOBundle: iobundle is nil")
+		return types.ClassETHER
+	}
+	switch ioBundle.Type {
```

Comment:
I don't think we should define classes but instead add a numeric priority in the API. For now you can use the freeUplink which is in DeviceNetworkStatus; later we'll add  the numeric priority end to end.